sql-plugin/src/main/scala/com/nvidia/spark/rapids/DeviceMemoryEventHandler.scala

```diff
@@ -30,53 +30,79 @@ import org.apache.spark.sql.rapids.execution.TrampolineUtil
  * RMM event handler to trigger spilling from the device memory store.
  * @param store device memory store that will be triggered to spill
  * @param oomDumpDir local directory to create heap dumps on GPU OOM
+ * @param isGdsSpillEnabled true if GDS is enabled for device->disk spill
+ * @param maxFailedOOMRetries maximum number of retries for OOMs after
+ *                            depleting the device store
  */
 class DeviceMemoryEventHandler(
     store: RapidsDeviceMemoryStore,
     oomDumpDir: Option[String],
-    isGdsSpillEnabled: Boolean) extends RmmEventHandler with Logging {
+    isGdsSpillEnabled: Boolean,
+    maxFailedOOMRetries: Int) extends RmmEventHandler with Logging with Arm {
 
   // Flag that ensures we dump stack traces once and not for every allocation
   // failure. The assumption is that unhandled allocations will be fatal
   // to the process at this stage, so we only need this once before we exit.
   private var dumpStackTracesOnFailureToHandleOOM = true
 
+  // Thread local used to the number of times a sync has been attempted, when
+  // handling OOMs after depleting the device store.
+  private val synchronizeAttempts = ThreadLocal.withInitial[Int](() => 0)
+
   /**
    * Handles RMM allocation failures by spilling buffers from device memory.
    * @param allocSize the byte amount that RMM failed to allocate
+   * @param retryCount the number of times this allocation has been retried after failure
    * @return true if allocation should be reattempted or false if it should fail
    */
-  override def onAllocFailure(allocSize: Long): Boolean = {
+  override def onAllocFailure(allocSize: Long, retryCount: Int): Boolean = {
     try {
-      val nvtx = new NvtxRange("onAllocFailure", NvtxColor.RED)
-      try {
+      withResource(new NvtxRange("onAllocFailure", NvtxColor.RED)) { _ =>
         val storeSize = store.currentSize
+        val attemptMsg = if (retryCount > 0) {
+          s"Attempt ${retryCount}. "
+        } else {
+          "First attempt. "
+        }
         logInfo(s"Device allocation of $allocSize bytes failed, device store has " +
-            s"$storeSize bytes. Total RMM allocated is ${Rmm.getTotalBytesAllocated} bytes.")
+            s"$storeSize bytes. $attemptMsg" +
```

Comment:
nit spacing is off

Reply:
Should be good here: https://github.com/NVIDIA/spark-rapids/pull/6849/commits/6d10d77081f28b950cde00775cace01e41313288